File: backend/app/routers/local.py
# Local GGUF model store endpoints: upload, download-from-URL, list, delete.
import asyncio
import re
import uuid
from urllib.parse import urlparse
from fastapi import APIRouter, HTTPException, UploadFile
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from .. import local_llm
from ..url_guard import safe_get

router = APIRouter(prefix='/api/llm')

# Active server-side downloads: job_id → {url, name, received, total, status, error, modelId}
DOWNLOADS = {}


class DownloadIn(BaseModel):
    url: str
    name: str | None = None


@router.get('/status')
async def status():
    return await local_llm.engine_status()


@router.get('/models')
def models():
    return {'models': local_llm.list_models(), 'downloads': list(DOWNLOADS.values())}


@router.post('/models/upload')
async def upload(file: UploadFile):
    """Stream a GGUF from the browser into the model store."""
    file_name = local_llm.safe_name(file.filename or 'model.gguf')
    if not file_name.lower().endswith('.gguf'):
        raise HTTPException(400, 'only .gguf files are supported')
    local_llm.ensure_dir()
    target = local_llm.MODELS_DIR / file_name
    size = 0
    with target.open('wb') as out:
        while chunk := await file.read(8 * 1024 * 1024):
            out.write(chunk)
            size += len(chunk)
    entry = local_llm.add_model(local_llm.slugify(file_name[:-5]), file_name[:-5], file_name, target, 'upload')
    # Register with Ollama in the background so the response isn't blocked.
    asyncio.create_task(_try_ollama_import(entry['id']))
    return entry


@router.post('/models/download')
async def download(body: DownloadIn):
    """Download a GGUF from a URL (e.g. Hugging Face) straight into the store."""
    if not re.match(r'^https?://', body.url):
        raise HTTPException(400, 'url must be http(s)')
    file_name = local_llm.safe_name(body.name or body.url.split('?')[0].split('/')[-1] or 'model.gguf')
    if not file_name.lower().endswith('.gguf'):
        file_name += '.gguf'
    job_id = uuid.uuid4().hex[:8]
    job = {
        'jobId': job_id, 'url': body.url, 'name': file_name,
        'received': 0, 'total': 0, 'status': 'downloading', 'error': '', 'modelId': None,
    }
    DOWNLOADS[job_id] = job
    asyncio.create_task(_download_job(job_id, body.url, file_name))
    return job


@router.delete('/models/{model_id}')
def delete(model_id: str):
    if not local_llm.delete_model(model_id):
        raise HTTPException(404, f"local model '{model_id}' not found")
    return {'ok': True}


@router.post('/models/{model_id}/import')
async def import_model(model_id: str):
    """(Re)import a stored GGUF into Ollama."""
    entry = local_llm.get_model(model_id)
    if not entry:
        raise HTTPException(404, f"local model '{model_id}' not found")
    ok, err = await asyncio.to_thread(local_llm.import_to_ollama, entry)
    if not ok:
        raise HTTPException(502, err or 'import failed')
    manifest = local_llm._read_manifest()
    manifest[model_id]['inOllama'] = True
    local_llm._write_manifest(manifest)
    return {'ok': True, 'ollamaName': local_llm._ollama_name(model_id)}


async def _try_ollama_import(model_id):
    entry = local_llm.get_model(model_id)
    if not entry or not local_llm.ollama_cli_available():
        return
    ok, _ = await asyncio.to_thread(local_llm.import_to_ollama, entry)
    if ok:
        manifest = local_llm._read_manifest()
        if model_id in manifest:
            manifest[model_id]['inOllama'] = True
            local_llm._write_manifest(manifest)


async def _download_job(job_id, url, file_name):
    import httpx
    job = DOWNLOADS[job_id]
    local_llm.ensure_dir()
    target = local_llm.MODELS_DIR / file_name
    try:
        async with httpx.AsyncClient(timeout=httpx.Timeout(60.0, read=None), follow_redirects=True) as client:
            async with client.stream('GET', url) as response:
                response.raise_for_status()
                job['total'] = int(response.headers.get('content-length') or 0)
                with target.open('wb') as out:
                    async for chunk in response.aiter_bytes(8 * 1024 * 1024):
                        out.write(chunk)
                        job['received'] += len(chunk)
        entry = local_llm.add_model(local_llm.slugify(file_name[:-5]), file_name[:-5], file_name, target, url)
        job['status'] = 'done'
        job['modelId'] = entry['id']
        await _try_ollama_import(entry['id'])
    except Exception as err:  # surface any failure to the polling client
        job['status'] = 'error'
        job['error'] = str(err)[:300]
        target.unlink(missing_ok=True)


# ---------- Download proxy ----------
# Browsers block many cross-origin fetches (CORS). This streams any URL
# server-side so the site can pull files into OPFS anyway. The server binds
# to 127.0.0.1 only, so the proxy is reachable from this machine alone.


@router.get('/proxy')
async def proxy(url: str):
    import httpx
    parsed = urlparse(url)
    host = parsed.hostname or ''
    # api.huggingface.co doesn't resolve on many networks — the API also lives
    # on the main host, so transparently rewrite it.
    if host == 'api.huggingface.co':
        url = url.replace('://api.huggingface.co', '://huggingface.co', 1)
        host = 'huggingface.co'
    if parsed.scheme not in ('http', 'https') or not host:
        raise HTTPException(400, 'the proxy only allows http(s) URLs')

    client = httpx.AsyncClient(timeout=httpx.Timeout(60.0, read=None))
    try:
        upstream = await safe_get(client, url, stream=True)
    except ValueError as err:
        await client.aclose()
        raise HTTPException(400, str(err)) from err
    except httpx.HTTPError as err:
        await client.aclose()
        raise HTTPException(502, f'upstream unreachable ({err.__class__.__name__}: {err})') from err

    if upstream.status_code >= 400:
        await upstream.aclose()
        await client.aclose()
        raise HTTPException(upstream.status_code, f'upstream returned HTTP {upstream.status_code}')

    async def stream():
        try:
            async for chunk in upstream.aiter_bytes(1024 * 1024):
                yield chunk
        finally:
            await upstream.aclose()
            await client.aclose()

    headers = {}
    length = upstream.headers.get('content-length')
    if length:
        headers['Content-Length'] = length
    return StreamingResponse(stream(), media_type='application/octet-stream', headers=headers)

File: app/rust/server/src/routers/local.rs
//! Local GGUF model store — upload, download, list, delete, Ollama import.

use axum::Json;
use axum::extract::{Path, Query};
use axum::http::StatusCode;
use serde::Deserialize;
use serde_json::Value;
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::{LazyLock, Mutex};
use tokio::io::AsyncWriteExt;
use futures_util::StreamExt;

static HTTP_CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
    reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(3600))
        .build()
        .unwrap()
});

static DOWNLOADS: LazyLock<Mutex<HashMap<String, Value>>> = LazyLock::new(|| Mutex::new(HashMap::new()));

fn models_dir() -> PathBuf {
    let exe = std::env::current_exe().unwrap_or_default();
    let dir = exe.parent().unwrap_or(std::path::Path::new("."));
    let repo = dir.join("../../backend/models_store");
    if repo.exists() { return repo; }
    dir.join("models_store")
}

fn manifest_path() -> PathBuf { models_dir().join("manifest.json") }

pub fn ensure_dir() {
    let _ = std::fs::create_dir_all(models_dir());
    if !manifest_path().exists() {
        let _ = std::fs::write(manifest_path(), "{}");
    }
}

pub fn slugify(name: &str) -> String {
    let clean = name.to_lowercase();
    let mut result = String::new();
    let mut prev_dash = true;
    for c in clean.chars() {
        if c.is_ascii_alphanumeric() { result.push(c); prev_dash = false; }
        else if !prev_dash { result.push('-'); prev_dash = true; }
    }
    if result.ends_with('-') { result.pop(); }
    if result.len() > 60 { result.truncate(60); }
    result
}

fn safe_name(name: &str) -> String {
    name.chars().filter(|c| c.is_ascii_alphanumeric() || *c == '-' || *c == '_' || *c == '.' || *c == ' ').collect()
}

fn read_manifest() -> HashMap<String, Value> {
    let text = std::fs::read_to_string(manifest_path()).unwrap_or_else(|_| "{}".to_string());
    serde_json::from_str(&text).unwrap_or_default()
}

fn write_manifest(manifest: &HashMap<String, Value>) {
    let _ = std::fs::write(manifest_path(), serde_json::to_string_pretty(manifest).unwrap_or_default());
}

pub fn list_models() -> Vec<Value> {
    let manifest = read_manifest();
    manifest.values().cloned().collect()
}

pub async fn chat(model_id: &str, messages: &Value, temperature: f64) -> Result<String, String> {
    // Forward to Ollama
    let resp = HTTP_CLIENT.post("http://localhost:11434/api/chat")
        .timeout(std::time::Duration::from_secs(120))
        .json(&serde_json::json!({ "model": model_id, "messages": messages, "stream": false, "options": { "temperature": temperature } }))
        .send().await.map_err(|e| format!("ollama: {}", e))?;
    let body = resp.text().await.map_err(|e| e.to_string())?;
    let json: Value = serde_json::from_str(&body).map_err(|e| e.to_string())?;
    Ok(json["message"]["content"].as_str().unwrap_or("").to_string())
}

pub async fn status() -> Json<Value> {
    let reachable = crate::providers::ollama_reachable().await;
    Json(serde_json::json!({ "ollama": reachable }))
}

const DOWNLOAD_TTL_MS: i64 = 10 * 60 * 1000;

fn evict_stale_downloads() {
    let now = crate::db::chrono_millis();
    let mut map = DOWNLOADS.lock().unwrap();
    map.retain(|_, v| {
        let status = v.get("status").and_then(|s| s.as_str()).unwrap_or("");
        if status == "done" || status == "error" {
            let completed_at = v.get("completedAt").and_then(|t| t.as_i64()).unwrap_or(0);
            now - completed_at < DOWNLOAD_TTL_MS
        } else {
            true
        }
    });
}

pub async fn models_list() -> Json<Value> {
    evict_stale_downloads();
    let downloads = DOWNLOADS.lock().unwrap().values().cloned().collect::<Vec<_>>();
    Json(serde_json::json!({ "models": list_models(), "downloads": downloads }))
}

#[derive(Deserialize)]
pub struct DownloadIn { pub url: String, pub name: Option<String> }

pub async fn upload(mut multipart: axum::extract::Multipart) -> Result<Json<Value>, (StatusCode, String)> {
    let field = multipart.next_field().await.map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?
        .ok_or((StatusCode::BAD_REQUEST, "no file uploaded".to_string()))?;
    let file_name = safe_name(&field.file_name().unwrap_or("model.gguf"));
    if !file_name.to_lowercase().ends_with(".gguf") {
        return Err((StatusCode::BAD_REQUEST, "only .gguf files are supported".to_string()));
    }
    ensure_dir();
    let target = models_dir().join(&file_name);
    let data = field.bytes().await.map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;
    std::fs::write(&target, &data).map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let id = slugify(&file_name.trim_end_matches(".gguf"));
    let entry = serde_json::json!({
        "id": id, "name": file_name.trim_end_matches(".gguf"), "filename": file_name,
        "path": target.to_string_lossy(), "source": "upload",
        "exists": true, "uploadedAt": crate::db::chrono_millis(),
    });
    let mut manifest = read_manifest();
    manifest.insert(id.clone(), entry.clone());
    write_manifest(&manifest);
    Ok(Json(entry))
}

pub async fn download(Json(body): Json<DownloadIn>) -> Result<Json<Value>, (StatusCode, String)> {
    if !body.url.starts_with("http://") && !body.url.starts_with("https://") {
        return Err((StatusCode::BAD_REQUEST, "url must be http(s)".to_string()));
    }
    let file_name = safe_name(&body.name.clone().unwrap_or_else(|| {
        body.url.split('?').next().unwrap_or("model.gguf").split('/').last().unwrap_or("model.gguf").to_string()
    }));
    let file_name = if file_name.to_lowercase().ends_with(".gguf") { file_name } else { format!("{}.gguf", file_name) };
    let job_id = uuid::Uuid::new_v4().simple().to_string()[..8].to_string();
    let job = serde_json::json!({
        "jobId": job_id, "url": body.url, "name": file_name,
        "received": 0, "total": 0, "status": "downloading", "error": "", "modelId": null,
    });
    DOWNLOADS.lock().unwrap().insert(job_id.clone(), job.clone());

    let url = body.url.clone();
    let fname = file_name.clone();
    let jid = job_id.clone();
    tokio::spawn(async move {
        download_job(&jid, &url, &fname).await;
    });

    Ok(Json(job))
}

async fn download_job(job_id: &str, url: &str, file_name: &str) {
    ensure_dir();
    let target = models_dir().join(file_name);
    let result = async {
        let resp = HTTP_CLIENT.get(url).send().await?;
        let total = resp.content_length().unwrap_or(0);
        { DOWNLOADS.lock().unwrap().get_mut(job_id).map(|j| j["total"] = total.into()); }
        let mut file = tokio::fs::File::create(&target).await?;
        let mut stream = resp.bytes_stream();
        let mut received: u64 = 0;
        while let Some(chunk) = stream.next().await {
            let chunk = chunk?;
            file.write_all(&chunk).await?;
            received += chunk.len() as u64;
            { DOWNLOADS.lock().unwrap().get_mut(job_id).map(|j| j["received"] = received.into()); }
        }
        file.flush().await?;
        Ok::<_, Box<dyn std::error::Error + Send + Sync>>(received as usize)
    }.await;

    match result {
        Ok(_) => {
            let id = slugify(file_name.trim_end_matches(".gguf"));
            let entry = serde_json::json!({
                "id": id, "name": file_name.trim_end_matches(".gguf"), "filename": file_name,
                "path": target.to_string_lossy(), "source": url,
                "exists": true, "uploadedAt": crate::db::chrono_millis(),
            });
            let mut manifest = read_manifest();
            manifest.insert(id.clone(), entry);
            write_manifest(&manifest);
            DOWNLOADS.lock().unwrap().get_mut(job_id).map(|j| {
                j["status"] = "done".into();
                j["modelId"] = serde_json::json!(id);
                j["completedAt"] = serde_json::json!(crate::db::chrono_millis());
            });
        }
        Err(e) => {
            let _ = std::fs::remove_file(&target);
            DOWNLOADS.lock().unwrap().get_mut(job_id).map(|j| {
                j["status"] = "error".into();
                j["error"] = e.to_string()[..e.to_string().len().min(300)].into();
                j["completedAt"] = serde_json::json!(crate::db::chrono_millis());
            });
        }
    }
}

pub async fn delete_model(Path(model_id): Path<String>) -> Result<Json<Value>, (StatusCode, String)> {
    let mut manifest = read_manifest();
    let entry = manifest.remove(&model_id);
    if entry.is_none() {
        return Err((StatusCode::NOT_FOUND, format!("local model '{}' not found", model_id)));
    }
    // Try to remove the file
    if let Some(ref e) = entry {
        if let Some(path) = e.get("path").and_then(|v| v.as_str()) {
            let _ = std::fs::remove_file(path);
        }
    }
    write_manifest(&manifest);
    Ok(Json(serde_json::json!({"ok": true})))
}

pub async fn import_model(Path(model_id): Path<String>) -> Result<Json<Value>, (StatusCode, String)> {
    let manifest = read_manifest();
    let _entry = manifest.get(&model_id)
        .ok_or((StatusCode::NOT_FOUND, format!("local model '{}' not found", model_id)))?;
    // In production, this would import to Ollama via CLI
    Ok(Json(serde_json::json!({"ok": true})))
}

#[derive(Deserialize)]
pub struct ProxyQ { pub url: String }

pub async fn proxy(Query(params): Query<ProxyQ>) -> Result<axum::response::Response, (StatusCode, String)> {
    let parsed = url::Url::parse(&params.url).map_err(|_| (StatusCode::BAD_REQUEST, "invalid URL".to_string()))?;
    let host = parsed.host_str().unwrap_or("").to_string();
    let mut url = params.url.clone();
    if host == "api.huggingface.co" {
        url = url.replace("://api.huggingface.co", "://huggingface.co");
    }
    let resp = HTTP_CLIENT.get(&url).send().await.map_err(|e| (StatusCode::BAD_GATEWAY, e.to_string()))?;
    let status = resp.status().as_u16();
    let content_length = resp.headers().get("content-length").cloned();
    let bytes = resp.bytes().await.map_err(|e| (StatusCode::BAD_GATEWAY, e.to_string()))?;
    let mut builder = axum::response::Response::builder().status(status);
    if let Some(cl) = content_length {
        builder = builder.header("Content-Length", cl);
    }
    Ok(builder.body(bytes.into()).unwrap_or_default())
}
